### ISSUE TITLE
Consider when to voice shape name - 3 types of context responses - Shape Name, Key Shape Properties, Shape Maintenance 

### ISSUE THREAD

**Issue by terracoda** [phetsims/quadrilateral#198]
Now with the show/hide shape name toggle interaction, we need to consider if we want to always voice the shape name or not.

I think if shape name is hidden, we would also not want to announce the shape name in any of the responsive descriptions, and possibly not in the state descriptions (i.e., Details or Hints button).

**Comment by terracoda** [comment]
Don't announce shape name when shape name is is hidden.

**Comment by terracoda** [comment]
When shape name is shown, and a transformation creates a named shape, we should hear:

- "Found a {{shapeName}}!" as the context response, instead of any context response of changing corners or sides.

The above context response, should replace any "Lost your parallelogram." and any "Found a parallelogram" context responses.

When shape name is not shown, it might be more tricky, but these are the shape properties I propose we call out:
- When a square is made: 
    - "All right angles and all sides equal." 
    - Replaces  "Found a parallelogram with all right angles and all sides equal." 

- When a rectangle is made:
    - "All right angles."
    - Replaces  "Found a parallelogram with all right angles."

- When a rhombus is made:
    - "All sides equal."
    - Replaces "Found a parallelogram with all sides equal."

- When a general parallelogram is made:
    - "Opposite sides tilt in parallel." (we already have this in a context response)
    - Replaces "Found a parallelogram."

Other shape transitions all currently trigger "Lost your parallelogram." as a context response. 
If we consider the origin shape of the transition, we might be able to have shorter context responses, but if we ignore that for now, we could have the following. 
- When a trapezoid is made:
    - "Opposite sides {{sideName}} and {{sideName}} tilt in parallel." (we already have this in a context response)
  
- When an isosceles trapezoid is made:
    - "Opposite sides {{sideName}} and {{sideName}} are equal, and Opposite sides {{sideName}} and {{sideName}} tilt in parallel." 

- When a kite or dart is made:
    - "Opposite corners {{cornerName}} and {{cornerName}} equal, pairs of adjacent sides are equal." 

- When a dart is made:
      - "{{cornerName}} points inward, Opposite corners {{cornerName}} and {{cornerName}} are equal, and pairs of adjacent sides are equal." 

- When convex quad is made:
    - "All sides tilt away from parallel." (we already have this in a context response)

- When concave quad is made:
    - "All sides tilt away from parallel, and {{cornerName}} points inward."

**Comment by terracoda** [comment]
@jessegreenberg, I'm still working on this.

**Comment by terracoda** [comment]
Again, when shape name is hidden, I think it would be better to just name the most important properties as they become true. That approach might mean that we not distinguish with a response between certain shapes, like Kites vs. Darts and Convex and Concave quadrilaterals, or perhaps Darts and Concave Quadrilaterals.

**Comment by BLFiedler** [comment]
Discussed over Zoom with TS, but a few notes I had:

I'm really for whatever feels most helpful at this point for clarity on what's happening with the shape. If that's describing the shape's appearances, what's equal or not equal, or what is or isn't parallel, then those are what's important.

Should we maintain the dynamic language in the context response for parallel when connecting a specific shape or treat it more in terms of what is and isn't true (and reserve "tilt" for case where a movement keeps us in a shape category)

**Comment by terracoda** [comment]
Ok, in that case, I would suggest that when shape name is hidden we call out properties as they become true. I don't want to have 2 different models - one for shape detect when shape name is visible and one for property detection.  I am pretty sure these 2 things are closely aligned, it is just to be a shape you several shape property test might need to be true.

I am wondering if this would make for two interesting exploration modes.

I am going to copy the test properties from #188

**Comment by terracoda** [comment]
We think this a good approach. We can voice properties as they happen when shape name is not shown, and voice the shape name when the shape name is shown.

I still need to figure out how these new context responses would be supported by the other responses we have built in currently.

**Comment by terracoda** [comment]
Ok, ignoring changes to object responses, @jessegreenberg, could you implement these context responses.

**Comment by terracoda** [comment]
The context responses in [commen](https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1228881882)

**Comment by terracoda** [comment]
@jessegreenberg, if you have categories for context responses in the code that affect prioritization, these context responses would be replacing the basic responses we discussed in https://github.com/phetsims/quadrilateral/issues/143#issue-1226816631

**Comment by jessegreenberg** [comment]
OK, this was done with the above commit. This ended up being a pretty big change, ~500 lines. But I think the design is sounding really nice @terracoda! Here is a dev version in case phettest has problems (big common code changes today make me think this is likely).

https://phet-dev.colorado.edu/html/quadrilateral/1.0.0-dev.62/phet/quadrilateral_en_phet.html

Over to you to review.

**Comment by terracoda** [comment]
I tested using my iPhone with only context responses and it sounded very good. More testing the morning!!

**Comment by terracoda** [comment]
When reviewing this issue yesterday I think the responses  sounded really good! And they still do.

I noticed there were no changing "object responses" firing during changes if context responses was unchecked. When I unchecked context responses I only get object responses when clicking on a side or corner. This is kind of a nice experience, in my opinion. So just noting it. I don't want any thing changed

**Design Action needed for a new issue:**
- Document the different responses for comparison.
- Document how the responses are prioritized.
- Figure out if any of them fall nicely into the object response category.
- Figure out which ones we don’t need/could live without.
- Then determine if anything is missing.
- Finally, determine if there are any timing adjustments needed.

Closing this issue.

**Comment by terracoda** [comment]
Apologies @jessegreenberg, I am re-opening as I would like to change the wording of one response.
When a Convex Quadrilateral is made, please use this wording:
- All sides out of parallel. 
    - Instead of, "All sides tilt away from parallel."

That change makes the transformative response more in line with the style of the other transformative responses.

Then, as a Convex Quadrilateral shape is maintained, the longer original response can be mixed in.

I am doing a review of all the responses over in #216 

I think you can close this issue when you make that change. Everything else sounded really good.

**Comment by terracoda** [comment]
@jessegreenberg ignore previous comment...

Actually 2 shape property responses (convex and concave) share that string, so best to update them both and I think this is better:

For Convex, instead of "All sides tilt away from parallel."
- No sides in parallel.

For Concave, instead of "Corner {{cornerLetter}} points inward, and all sides tilt away from parallel."
- Corner {{cornerLetter}} points inward, no sides in parallel.

Note that I dropped the "and".

Please keep the other phrases around:
- "All sides tilt away from parallel."
- "Corner {{cornerLetter}} points inward, and all sides tilt away from parallel."

They might be helpful for maintaining convex and concave shape properties.

**Comment by jessegreenberg** [comment]
OK above comment done in the above commit.

> Please keep the other phrases around:
> ...
> "Corner {{cornerLetter}} points inward, and all sides tilt away from parallel."

I am confused, this was requested to be replaced. I don't see this string used anywhere anymore. Is that correct? @terracoda over to you to review.

**Comment by terracoda** [comment]
In https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1228881882 I requested 

> When concave quad is made:
> 
>     " Corner {{cornerLetter}} points inward, and all sides tilt away from parallel."

So was just asking for that string to be updated.

**Comment by terracoda** [comment]
Still working on responses, there seem to be 3 types categories of shape context responses:
1. When Shape Name is shown: announcing "Found a {{ShapeName}}!"
2. When Shape Name is not shown: announcing the most defining properties of shape (see https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1228881882)
3. When maintaining a shape and the side lengths or angles change enough to trigger other context responses.

In #216 I am reviewing the responses to see how they fit together. I think most upcoming changes will be to object responses. And adjustments to the comparative scales are needed.

**Comment by jessegreenberg** [comment]
For the confusion in https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1247178579

@terracoda and I met over zoom and she said that she wanted to keep the strings like
```
      "allSidesTiltAwayFromParallelAsShapeChangesPattern": {
        "value": "All sides tilt away from parallel as shape gets {{areaChange}}"
      },
```

They are still in the file if we need.

**Comment by terracoda** [comment]
Requesting a few more tweaks to the strings that describe the key shape properties when Shape Name is hidden.

When making a Trapezoid:
- Change: Side {{sideLetter}} and Side {{sideLetter}} in parallel.
- To: Opposite sides {{sideLetter}} and {{sideLetter}} in parallel.

When making an Isosceles Trapezoid:
- Change: Side {{sideLetters}} and Side {{sideLetters}} are equal, and sides {{sideLetter}} and {{sideLetter}} in parallel.
- To: Opposite sides {{sideLetter}} and {{sideLetters}} are equal, and opposite sides {{sideLetter}} and {{sideLetter}} in parallel.

When Making a Square:
- Change: All right angles and all sides equal.
- To: All right angles, all sides equal.

That should do it for these context responses.

**Comment by terracoda** [comment]
My apologies @jessegreenberg, I forgot to re-assign you.

**Comment by terracoda** [comment]
Shape Name Context Response | Shape Property Context Response | Responses for Maintaining
-------------------------------- | ----------------------------------- | --------------------------
| Convex Quadrilateral | No sides in parallel. | All sides tilt away from parallel as shape gets {{bigger/smaller}}. |
| Concave Quadrilateral | Corner {{cornerLetter}} points inward, no sides in parallel. | Same as  Convex Quad for now. |
| Dart | Corner {{cornerLetter}} points inward, Corner {{cornerLetter}} and Corner {{cornerLetter}} are equal, and pairs of adjacent sides are equal. | Sides {{sideLetter}} and {{sideLetters}} change {{equally}}, as shape gets {{bigger/smaller}}. |
| Kite | Opposite corners {{cornerLetter}} and {{cornerLetter}} are equal, and pairs of adjacent sides are equal. | Same response as for maintaining a Dart for now. |
| Trapezoid | Opposite sides {{sideLetter}} and {{sideLetter}} in parallel. | Opposite sides {{SideLetter}} and {{SideLetter}} tilt in parallel as shape gets {{bigger/smaller}}. |
| Isosceles Trapezoid | Opposite sides {{sideLetter}} and {{sideLetters}} equal, opposite sides {{sideLetter}} and {{sideLetter}} in parallel. | Opposite sides {{sideLetter}} and {{sideLetters}} {{equal}}, as shape gets {{bigger/smaller}}. |
| Parallelogram | Opposite sides in parallel. | Opposite sides {{tilt/straighten}} in parallel as opposite corners change {{equally}}. Opposite sides in parallel as shape gets {{bigger/smaller}}. |
| Rectangle | All right angles. | {{All right angles}} as shape gets {{bigger/smaller}}. |
| Rhombus | All sides equal. | Opposite corners change {{equally}} as {{equal}} sides {{tilt/straighten}} in parallel. |
| Square | All right angles, all sides equal. | {{All sides}} equal as shape gets {{bigger/smaller}}. |

**Comment by terracoda** [comment]
@jessegreenberg, the maintaining context responses won't need to fire with every key press, I'll need to give you some basic angle and side length change thresholds.

**Comment by terracoda** [comment]
I italicized the changed responses in https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1256381989

**Comment by terracoda** [comment]
@jessegreenberg, is it possible to similar thresholds/timing that are currently implemented for maintaining the parallelogram shapes, and the quadrilateral that are not parallelograms?

We might have to tweak, but maybe those are good starting points?

**Comment by terracoda** [comment]
So for example for Parallelograms:
- Rectangle: as long as I remain a Rectangle (all angles equal) say my maintenance response as often as we hear the area-scaling response for parallelograms.
- Square: as long as I remain a Square (all sides equal & all corners equal) say my maintenance response as often as we hear the area-scaling response for parallelograms.
- Rhombus: as long as I remain a Rhombus (all sides equal) say my maintenance response as often as we hear the angle changing response for parallelograms (non-area-scaling response).

For trapezoids, concave quads, convex quads, kites and darts do we we already have some area-scaling thresholds in place?

**Comment by terracoda** [comment]
@jessegreenberg, if we need to discuss/adjust the timing of the "Responses for Maintaining Shapes" I will create a new issue!

**Comment by jessegreenberg** [comment]
The above commit updated to the latest design for "shape context response (shape name hidden)". I think I understand what you mean in https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1256418584 @terracoda, yes Ill try that.

**Comment by terracoda** [comment]
The Shape Maintenance Responses should fire regardless of whether or not Shape Name is hidden. I am not sure I made that clear (sorry).

The ways to maintain either a Square or Rhombus are quite limited. For example, for Rhombus, I have not been able to maintain side length while trying to make the arc movement. SO, let's ignore those 2 shapes as we try to sort out the other responses for "maintaining a shape".

I don't yet hear the shape maintenance context responses for subsets of parallelograms and trapezoids (i.e., the italicized ones in https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1256381989). Maybe the initial thresholds aren't working for these responses? I still hear the original responses.

When **maintaining a rectangle** I want to hear: 
- {{All right angles}} as shape gets {{bigger/smaller}}.
- But I still hear the parallelogram maintenance response.

When **maintaining a Isosceles Trapezoid** I want to hear: 
- Opposite sides {{sideLetter}} and {{sideLetters}} {{equal}}, as shape gets {{bigger/smaller}}.
- But I still hear the regular trapezoid maintenance response.

When **maintaining a Dart or Kite** I want to hear: 
- Sides {{sideLetter}} and {{sideLetters}} change {{equally}}, as shape gets {{bigger/smaller}}.
- But I still hear the regular convex/concave quad (non-parallelogram) maintenance response.

I think we need to look at the logic and tolerances for the 3 groups of shapes mentioned above.

The design idea for the category of Shape-Maintaining Context Responses  is for them to fire occasionally as the shape continues to maintain its own set of true-ness properties. So in theory when maintaining a shape, the shape has to be true for itself first before it can be maintained. 

When shapes are being recognized correctly by the model, I would expect to hear context responses like this:

1. When a shape becomes true, I would expect to hear the Shape Property Response or Shape Name Response. This depends on the state of the Show Shape toggle. This is currently working well.
2. Then after a bit of movement with no change to the true-ness of the shape properties, I would expect to hear the Shape Maintenance Response for the current shape.
3.  When the shape changes to a new shape, I would then hear item 1 again.
4. If an interaction begins in a shape and the shape is maintained, I would NOT expect to hear item 1 until a different Shape becomes true.

I'm realizing that to maintain certain shapes, a learner is often restricted to moving either a side or a corner, so I might need to map that out. For example:
- To maintain a kite or dart, a learner must move a corner along the diagonal which is most easily accomplished with the keyboard when the diagonal is parallel to the grid.
- To maintain a rectangle, a learner must move a side without changing any angles which is easiest to do with the keyboard.
- To maintain a Convex/or Concave Quadrilateral, a learner can move either a corner or side.

With the imprecision of mouse input in the first 2 example cases, I would expect to hear more back and forth of item 1 (Shape Property or Shape Name Responses) and rarely hear the Shape Maintenance Response.

**Comment by BLFiedler** [comment]
> The ways to maintain either a Square or Rhombus are quite limited. For example, for Rhombus, I have not been able to maintain side length while trying to make the arc movement. SO, let's ignore those 2 shapes as we try to sort out the other responses for "maintaining a shape".

This is unlikely to change unless the tolerances are larger than the input spacing, which is unlikely to happen for the published version. It may be an option for research purposes to investigate maintaining shapes, but it significantly muddles mathematical accuracy that we've concerned ourselves with quite a bit at this point.

**Comment by terracoda** [comment]
@BLFiedler, that leads me to believe that we might only hear Square and Rhombus Maintenance Responses with the TMQ. 

Does that seem reasonable to you?

**Comment by jessegreenberg** [comment]
Sorry you reviewd this! I finished "shape property context response" but only got about 25% of the way through the "shape maintenance responses".  I think I understand what you are going for but it isnt implemented yet.

**Comment by terracoda** [comment]
Thanks for the update @jessegreenberg, apologies for jumping in too early. I'll look at the assignment label next time :-)

**Comment by terracoda** [comment]
@jessegreenberg, I found a case for a Convex Quad where we would need a different Maintenance Response. 
<img width="499" alt="image" src="https://user-images.githubusercontent.com/1902116/194410293-0fb1f148-dae7-4438-9c90-c8793260635b.png">

When moving Corner C in this configuration the Quad can be maintained without a change in area (size). This shape can also transform from Convex Quad to Kite without changing area. The transformation responses sound fine as they do not mention area.

Response coming soon.

**Comment by terracoda** [comment]
Ok, here's the side changing response for maintaining a convex quad with a 180 degree angle
-   {{CornerName}} {{flat}}, {{SideName}} gets {{longer/shorter}} and {{SideName}} gets {{longer/shorter}}.

**Comment by terracoda** [comment]
@jessegreenberg, I added the convex quad side-changing response to the table in https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1256381989, but it is dependent on the decision in #236

**Comment by terracoda** [comment]
Might need "adjacent" in the response - adding.

**Comment by jessegreenberg** [comment]
OK @terracoda this is mostly done with the following exceptions that would be good to discuss:
- Regarding 
  > here's the side changing response for maintaining a convex quad with a 180 degree angle...We can use one with fewer parameters if that makes sense in the code.

  Thanks! I used `{{CornerName}} {{flat}} as adjacent sides get change in length.` for now, it saved me ~30 minutes in the implementation to determine which sides are changing and in what direction.
- Square and Rhombus maintenance responses are implemented but I couldn't figure out a way to test them (at least with mouse/touch/keyboard).
- For the dart/kite responses
  > Sides {{sideLetter}} and {{sideLetters}} change {{equally}}, as shape gets {{bigger/smaller}}.
  
  I can't think of a way to find the two `sideLetter` parameters. What if a kite is maintained by pulling two opposite vertices apart? Is there a way we can determine this geometrically, or is there a different description we can use?

**Comment by jessegreenberg** [comment]
> Sides {{sideLetter}} and {{sideLetters}} change {{equally}}, as shape gets {{bigger/smaller}}.

Discussed in a design meeting today, lets go with "Adjacent sides change equally as shape gets {{bigger/smaller}}.

**Comment by jessegreenberg** [comment]
OK, that update is done. This is ready for your review @terracoda.

**Comment by terracoda** [comment]
All Shape Name Context Responses seem to work just fine!
All Shape Property Context Responses seem to work just fine, though I might need a couple tweaks to wording.

Some things about the Shape Maintenance Context Responses surprised me. Here's what I found surprising, and need or might need adjusting:
1. When maintaining a Parallelogram, I get two responses. One about the changing area and one about the corners changing equally. This is my fault (bad documentation above). We used to have 2 thresholds, one for each of these responses. 
    - Can we bring those thresholds back so the responses fire separately?
    - Also, the {{tilt/straighten}} parameter makes more sense when using the TMQ and less sense when using the virtual sim because when moving a side it is the sides that are adjacent to the moving side that straighten and tilt. I think we should remove "tilt/straighten" parameters and just have "in parallel". If we need a description of tilt for parallelograms or trapezoids, maybe that detail should be part of side object responses.
    - Let's remove the tilt/straighten parameter from all parallelograms
    - New separate maintenance response parallelograms when corners are changing more than the sides:
        - "Opposite sides in parallel as opposite corners change {{equally}}." 
 
2. When maintaining a regular Trapezoid by moving a corner or a side, though the angles change, the parallel sides don't actually change their tilt in any way. They remain in parallel. Let's also remove "tilt" from that trapezoid responses.
    - New Response for maintaining a trapezoid: 
        - "Opposite sides {{SideLetter}} and {{SideLetter}} in parallel as shape gets {{bigger/smaller}}."
        - No change to the response for changing area, "Opposite sides in parallel as shape gets {{bigger/smaller}}." We just need to re-implement the thresholds for these 2 responses.

3. Sometimes when maintaining an Isosceles Trapezoid, I felt like I wasn't hearing changes very often, but I think this is because I was listening with out Object Responses. No action required. I need to hear the responses with the new Object Responses when those are ready.

4. You already noticed this - I can maintain a Rhombus sometimes for about 3 key presses, and do not hear a maintenance response. I am wondering if we can adjust things so we get at least one response with keyboard input. **We can discuss.** This is not a show stopper. It will be more apparent with TMQ or other special inputs, or bi-manual input. Also, I might need to re-word the responses for maintaining a Rhombus.

5. Just an FYI, I think "tilt" still works when "all sides tilt away from parallel" when maintaining convex and concave quadrilaterals.

**Comment by terracoda** [comment]
@jessegreenberg, hopefully these are minor updates :-)

**Comment by jessegreenberg** [comment]
OK requests were done in the above commit. I tried to turn the comment into a checklist to identify action items. Things left unchecked had the following questions:

> Can we bring those thresholds back so the responses fire separately?

Yes, done - but only for parallelogram if I understand correctly.

> Let's remove the tilt/straighten parameter from all parallelograms

Tilt/straighten was removed from parallelogram maintenance responses. Can you give me an example of another response where tilt/straighten should be removed? There are so many strings and cases now I am having a hard time keeping track.

> I am wondering if we can adjust things so we hear one response with keyboard input.

The trigger for all context responses is
- If an angle has changed more than Math.PI / 24 OR
- If a length has changed more than a side segment length OR
- If we go from zero parallel sides to at least one pair of parallel sides

We could change these but it would change the trigger for ALL context responses. We could have one context response per key press but that seems very verbose.

Back to @terracoda!

**Comment by terracoda** [comment]
According to your comments above, @jessegreenberg, I think you likely got all the "tilts". I still need to review.

**Comment by jessegreenberg** [comment]
Thanks @terracoda, good find. This is another precision error, when moving the vertex back and forth I see that the area changes by values like 1.3322676295501878-15. Should be fixed in the above commit.

> Which response did you implement from https://github.com/phetsims/quadrilateral/issues/198#issuecomment-1271634080 ?

I went with "{{CornerName}} {{flat}} as adjacent sides get change in length."

Back to @terracoda to try it out.

**Comment by terracoda** [comment]
You fixed it. I am not hearing area changes anymore.
Thanks!

I think this issue is ready to close.
I'll think about when we should hear our fun question about the number of sides over in #236

**Comment by terracoda** [comment]
"{{CornerName}} {{flat}} as adjacent sides get change in length."

Should have been
"{{CornerName}} {{flat}} as adjacent sides change in length." with no "get"
